File: AndroidWebBox/coconut-core/src/main/java/com/sniper/coconut/bridge/BridgePerformance.kt
package com.sniper.coconut.bridge

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Bridge Performance Collector
 *
 * Lightweight singleton in coconut-core to record bridge call metrics.
 * PerformanceComponent in coconut-plugins reads from this.
 */
object BridgePerformance {

    data class CallRecord(
        val method: String,
        val durationMs: Long,
        val success: Boolean,
        val timestamp: Long = System.currentTimeMillis()
    )

    data class MethodStats(
        val method: String,
        var callCount: Long = 0,
        var successCount: Long = 0,
        var failCount: Long = 0,
        var totalDuration: Long = 0,
        var minDuration: Long = Long.MAX_VALUE,
        var maxDuration: Long = 0
    ) {
        val avgDuration: Long get() = if (callCount > 0) totalDuration / callCount else 0
        val successRate: Double get() = if (callCount > 0) successCount.toDouble() / callCount else 0.0
    }

    private val callHistory = ConcurrentHashMap<String, MutableList<CallRecord>>()
    private val methodStatsMap = ConcurrentHashMap<String, MethodStats>()
    private val totalCalls = AtomicLong(0)
    private val totalSuccess = AtomicLong(0)
    @Volatile
    var startTime: Long = System.currentTimeMillis()
        private set

    private const val MAX_HISTORY_PER_METHOD = 100

    fun record(method: String, durationMs: Long, success: Boolean) {
        totalCalls.incrementAndGet()
        if (success) totalSuccess.incrementAndGet()

        methodStatsMap.compute(method) { _, existing ->
            val stats = existing ?: MethodStats(method)
            stats.callCount++
            if (success) stats.successCount++ else stats.failCount++
            stats.totalDuration += durationMs
            stats.minDuration = minOf(stats.minDuration, durationMs)
            stats.maxDuration = maxOf(stats.maxDuration, durationMs)
            stats
        }

        callHistory.compute(method) { _, list ->
            val l = list ?: mutableListOf()
            l.add(CallRecord(method, durationMs, success))
            if (l.size > MAX_HISTORY_PER_METHOD) l.removeAt(0)
            l
        }
    }

    fun getTotalCalls(): Long = totalCalls.get()
    fun getTotalSuccess(): Long = totalSuccess.get()
    fun getMethodStatsMap(): Map<String, MethodStats> = methodStatsMap.toMap()
    fun getCallHistory(): Map<String, List<CallRecord>> = callHistory.mapValues { it.value.toList() }

    fun resetAll() {
        callHistory.clear()
        methodStatsMap.clear()
        totalCalls.set(0)
        totalSuccess.set(0)
        startTime = System.currentTimeMillis()
    }
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Bridge/BridgePerformance.swift
import Foundation

public class BridgePerformance {

    public static let shared = BridgePerformance()

    public struct CallRecord {
        public let method: String
        public let durationMs: Int64
        public let success: Bool
        public let timestamp: Int64
    }

    public struct MethodStats {
        public var callCount: Int = 0
        public var successCount: Int = 0
        public var failCount: Int = 0
        public var totalDuration: Int64 = 0
        public var minDuration: Int64 = Int64.max
        public var maxDuration: Int64 = 0

        public var avgDuration: Int64 {
            callCount > 0 ? totalDuration / Int64(callCount) : 0
        }
        public var successRate: Double {
            callCount > 0 ? Double(successCount) / Double(callCount) : 0
        }

        public func toDict() -> [String: Any] {
            return [
                "callCount": callCount,
                "successCount": successCount,
                "failCount": failCount,
                "totalDuration": totalDuration,
                "minDuration": minDuration == Int64.max ? 0 : minDuration,
                "maxDuration": maxDuration,
                "avgDuration": avgDuration,
                "successRate": successRate
            ]
        }
    }

    private var callHistory: [CallRecord] = []
    private var methodStatsMap: [String: MethodStats] = [:]
    private let lock = NSLock()

    private init() {}

    public func record(method: String, durationMs: Int64, success: Bool) {
        lock.lock()
        defer { lock.unlock() }

        let record = CallRecord(
            method: method,
            durationMs: durationMs,
            success: success,
            timestamp: Int64(Date().timeIntervalSince1970 * 1000)
        )
        callHistory.append(record)

        var stats = methodStatsMap[method] ?? MethodStats()
        stats.callCount += 1
        if success { stats.successCount += 1 } else { stats.failCount += 1 }
        stats.totalDuration += durationMs
        stats.minDuration = min(stats.minDuration, durationMs)
        stats.maxDuration = max(stats.maxDuration, durationMs)
        methodStatsMap[method] = stats
    }

    public func getTotalCalls() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return callHistory.count
    }

    public func getTotalSuccess() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return callHistory.filter { $0.success }.count
    }

    public func getMethodStatsMap() -> [String: MethodStats] {
        lock.lock()
        defer { lock.unlock() }
        return methodStatsMap
    }

    public func getCallHistory(limit: Int = 100) -> [CallRecord] {
        lock.lock()
        defer { lock.unlock() }
        return Array(callHistory.suffix(limit))
    }

    public func resetAll() {
        lock.lock()
        defer { lock.unlock() }
        callHistory.removeAll()
        methodStatsMap.removeAll()
    }
}
